Unable to basic auth using credentials option

Hi, 

### Version report

Jenkins and plugins versions report:

```
Jenkins version: 2.289.2-cb-3

& cat plugins.yaml 
plugins:
  - artifactId: git
    source:
      version: latest
```

I use the managed master docker image cloudbees/cloudbees-core-mm:2.289.2.2
In a container of the image :

```
$ uname -a
Linux 69d0e0fd597f 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

$ java -version
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (build 1.8.0_292-b10)
OpenJDK 64-Bit Server VM (build 25.292-b10, mixed mode)
```

### Reproduction steps

<!--
- Write bullet-point reproduction steps.
- Be explicit about any relevant configuration, jobs, build history, user accounts, etc., redacting confidential information as needed.
- The best reproduction steps start with a clean Jenkins install, perhaps a `docker run` command if possible.
- Use screenshots where appropriate, copy textual output otherwise. When in doubt, do both.
- Include relevant logs, debug if needed - https://www.jenkins.io/doc/book/system-administration/viewing-logs/
-->

I use jenkins-plugin-manager in a Dockerfile to install some plugins
We are proxifing Jenkins update center using Artifactory. Artifactory needs the requests to be identified with creds (basic auth user:password)

However I am unable to authenticate against Artifactory using the '--credentials' option

```
& export JENKINS_UC=https://repo.artifactory.lan/artifactory/cloudbees-update-center/update-center/cje-mm/update-center.json

& java -jar jenkins-plugin-manager-2.10.0.jar --plugin-file plugins.yaml  --verbose --credentials 'repo.artifactory.lan:username:password'

```

### Results

The jenkins-plugin-manager jar gets a 401 HTTP Error from the server :
```
& java -jar jenkins-plugin-manager-2.10.0.jar --plugin-file plugins.yaml  --verbose --credentials 'repo.artifactory.lan:username:password'
File containing list of plugins to be downloaded: plugins.yaml
Reading in plugins from plugins.yaml

No directory to download plugins entered. Will use default of /usr/share/jenkins/ref/plugins
Using update center https://repo.artifactory.lan/artifactory/cloudbees-update-center/update-center/cje-mm/update-center.json from JENKINS_UC environment variable
No CLI option or environment variable set for experimental update center, using default of https://updates.jenkins.io/experimental/update-center.json
No CLI option or environment variable set for incrementals mirror, using default of https://repo.jenkins-ci.org/incrementals
No CLI option or environment variable set for plugin info, using default of https://updates.jenkins.io/plugin-versions.json
No war entered. Will use default of /usr/share/jenkins/jenkins.war
Jenkins version: 2.289.2-cb-3

Retrieving update center information
Cache miss for: update-center-2.289.2-cb-3
io.jenkins.tools.pluginmanager.impl.UpdateCenterInfoRetrievalException: Error getting update center json
        at io.jenkins.tools.pluginmanager.impl.PluginManager.getJson(PluginManager.java:786)
        at io.jenkins.tools.pluginmanager.impl.PluginManager.getUCJson(PluginManager.java:805)
        at io.jenkins.tools.pluginmanager.impl.PluginManager.start(PluginManager.java:203)
        at io.jenkins.tools.pluginmanager.impl.PluginManager.start(PluginManager.java:167)
        at io.jenkins.tools.pluginmanager.cli.Main.main(Main.java:70)
Caused by: java.io.IOException: Server returned HTTP response code: 401 for URL: https://repo.artifactory.lan/artifactory/cloudbees-update-center/update-center/cje-mm/update-center.json?version=2.289.2-cb-3
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1900)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1498)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:268)
        at java.net.URL.openStream(URL.java:1068)
        at org.apache.commons.io.IOUtils.toString(IOUtils.java:2960)
        at io.jenkins.tools.pluginmanager.impl.PluginManager.getJson(PluginManager.java:780)
        ... 4 more
Error getting update center json
```

Artifactory logs say that the request is not authenticated :
```
non_authenticated_user|GET|/cloudbees-update-center/update-center/cje-mm/update-center.json|401|-1|0|3|Java/1.8.0_292
```

Expected result:

<!-- What was your expected result? -->
I would expect that the jar retrieves the update-center.json file from Artifactory just like I get it using a curl command with the same creds

```
& curl -I -u username:password https://repo.artifactory.lan/artifactory/cloudbees-update-center/update-center/cje-mm/update-center.json?version=2.289.2-cb-3
HTTP/1.1 200 OK
Date: Fri, 09 Jul 2021 20:47:43 GMT
Content-Type: application/json
Content-Length: 1856377
Connection: keep-alive
Last-Modified: Fri, 09 Jul 2021 17:08:29 GMT
ETag: 498d57992f13c805a953f487a35ef5d59a2ab671
X-Checksum-Sha1: 498d57992f13c805a953f487a35ef5d59a2ab671
X-Checksum-Sha256: b0870486193cb151e20b20f033f1ea7c8480f1c8a55460f38a767810b6ad074a
X-Checksum-Md5: d1afee726cc7a04896eccddf9b9fc170
Accept-Ranges: bytes
X-Artifactory-Filename: update-center.json
Content-Disposition: attachment; filename="update-center.json"; filename*=UTF-8''update-center.json
```

Am I missing something (in the credentials option host format ? tried a lot of combinations) if so could you please provide me with an example (couldn't find it)... or is this a bug ?

Thank you for your help :)

--- a/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java
+++ b/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java
@@ -46,6 +46,7 @@ import java.util.Properties;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ForkJoinPool;
+import java.util.function.Function;
 import java.util.jar.JarFile;
 import java.util.regex.Matcher;
 import java.util.stream.Collectors;
@@ -60,10 +61,12 @@ import org.apache.http.auth.AuthScope;
 import org.apache.http.auth.UsernamePasswordCredentials;
 import org.apache.http.client.CredentialsProvider;
 import org.apache.http.client.HttpClient;
+import org.apache.http.client.ResponseHandler;
 import org.apache.http.client.methods.HttpGet;
 import org.apache.http.client.protocol.HttpClientContext;
 import org.apache.http.client.utils.URIBuilder;
 import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.BasicResponseHandler;
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.DefaultHttpRequestRetryHandler;
 import org.apache.http.impl.client.HttpClients;
@@ -781,10 +784,19 @@ public class PluginManager implements Closeable {
                 System.out.println("Cache miss for: " + cacheKey);
             }
         }
-
+        final String response;
         try {
-            String urlText = IOUtils.toString(url, StandardCharsets.UTF_8);
-            String result = removePossibleWrapperText(urlText);
+            if (url.getProtocol().equalsIgnoreCase("http") || url.getProtocol().equalsIgnoreCase("https")) {
+                response = getViaHttpWithResponseHandler(
+                  url.toString(),
+                  new BasicResponseHandler(),
+                  cacheKey,
+                  e -> String.format("Unable to retrieve JSON from %s: %s", url, e.getMessage()),
+                  3);
+            } else {
+                response = IOUtils.toString(url, StandardCharsets.UTF_8);
+            }
+            String result = removePossibleWrapperText(response);
             JSONObject json = new JSONObject(result);
             cm.addToCache(cacheKey, json);
             return json;
@@ -899,7 +911,7 @@ public class PluginManager implements Closeable {
                     String.format("%nResolving dependencies of %s by downloading plugin to temp file %s and parsing " +
                             "MANIFEST.MF", plugin.getName(), tempFile.toString()));
             if (!downloadPlugin(plugin, tempFile)) {
-                Files.delete(tempFile.toPath());
+                Files.deleteIfExists(tempFile.toPath());
                 throw new DownloadPluginException("Unable to resolve dependencies for " + plugin.getName());
             }
 
@@ -1278,41 +1290,59 @@ public class PluginManager implements Closeable {
      * @param maxRetries   Maximum number of times to retry the download before failing
      * @return              boolean signifying if plugin was successfully downloaded
      */
-    @SuppressFBWarnings({"HTTP_PARAMETER_POLLUTION"})
     protected boolean downloadHttpToFile(String pluginUrl, Plugin plugin, File pluginFile, int maxRetries){
+        try {
+            getViaHttpWithResponseHandler(pluginUrl, new FileDownloadResponseHandler(pluginFile), plugin.getName(),
+                    e -> String.format("Unable to resolve plugin URL %s, or download plugin %s to file: %s", pluginUrl,
+                            plugin.getName(), e.getMessage()),
+                    maxRetries);
+
+            // Check if plugin is a proper ZIP file
+            try (JarFile ignored = new JarFile(pluginFile)) {
+                plugin.setFile(pluginFile);
+                logVerbose("Downloaded and validated plugin " + plugin.getName());
+            } catch (IOException e) {
+                throw new IOException("Downloaded file for " + plugin.getName() + " is not a valid ZIP", e);
+            }
+        } catch (IOException e) {
+            System.out.println(e.getMessage());
+            if (verbose) {
+                e.printStackTrace();
+            }
+            return false;
+        }
+        return true;
+    }
+
+    @SuppressFBWarnings({"HTTP_PARAMETER_POLLUTION"})
+    private <T> T getViaHttpWithResponseHandler(String url, ResponseHandler<? extends T> responseHandler, String resourceName, Function<IOException, String> ioExceptionMessageSupplier, int maxRetries) throws IOException {
+        HttpClient httpClient = getHttpClient();
+        HttpClientContext context = HttpClientContext.create();
+        CredentialsProvider credentialsProvider = getCredentialsProvider();
+        if (credentialsProvider != null) {
+            context.setCredentialsProvider(credentialsProvider);
+        }
+        HttpGet httpGet = new HttpGet(url);
         boolean success = false;
+        // TODO: retry logic should rather be implemented via DefaultHttpRequestRetruHandler, there is no need for an additional retry
         for (int i = 0; i < maxRetries; i++) {
-            success = true;
             try {
-                if (pluginFile.exists()) {
-                    Files.delete(pluginFile.toPath());
-                }
+                T response = httpClient.execute(httpGet, responseHandler, context);
+                return response;
             } catch (IOException e) {
-                logVerbose(String.format("Unable to delete %s before retry %d", pluginFile, i + 1));
-            }
-            HttpClient httpClient = getHttpClient();
-            HttpClientContext context = HttpClientContext.create();
-            CredentialsProvider credentialsProvider = getCredentialsProvider();
-            if (credentialsProvider != null) {
-                context.setCredentialsProvider(credentialsProvider);
-            }
-            HttpGet httpGet = new HttpGet(pluginUrl);
-            try {
-                httpClient.execute(httpGet, new FileDownloadResponseHandler(pluginFile), context);
-            } catch (IOException e) {
-                String message = String.format("Unable to resolve plugin URL %s, or download plugin %s to file: %s",
-                pluginUrl, plugin.getName(), e.getMessage());
-                if (i >= maxRetries -1) {
+                String message = ioExceptionMessageSupplier.apply(e);
+                if (i < maxRetries - 1) {
                     System.out.println(message);
                 } else {
-                    logVerbose(message);
+                    throw new IOException(message, e);
                 }
-                success = false;
             } finally {
                 // get final URI (after all redirects)
                 List<URI> locations = context.getRedirectLocations();
                 if (locations != null) {
-                    String message = String.format("%s %s from %s (attempt %d of %d)", success ? "Downloaded" : "Tried downloading", plugin.getName(), locations.get(locations.size() - 1), i+1, maxRetries);
+                    String message = String.format("%s %s from %s (attempt %d of %d)",
+                            success ? "Downloaded" : "Tried downloading", resourceName,
+                            locations.get(locations.size() - 1), i + 1, maxRetries);
                     if (success) {
                         logVerbose(message);
                     } else {
@@ -1320,28 +1350,11 @@ public class PluginManager implements Closeable {
                     }
                 }
             }
-
-            // Check if plugin is a proper ZIP file
-            if (success) {
-                try (JarFile ignored = new JarFile(pluginFile)) {
-                    plugin.setFile(pluginFile);
-                    logVerbose("Downloaded and validated plugin " + plugin.getName());
-                    break;
-                } catch (IOException e) {
-                    System.out.println("Downloaded file for " + plugin.getName() + " is not a valid ZIP");
-                    if (i >= maxRetries -1) {
-                        if (verbose) {
-                            e.printStackTrace();
-                        }
-                    }
-                    success = false;
-                }
-            }
         }
-        return success;
+        throw new IllegalStateException("Reached maximum number of retries without triggering IOException");
     }
 
-      /**
+    /**
      * Downloads a plugin from local folder location
      *
      * @param pluginUrl location to download plugin to.

--- a/plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerTest.java
+++ b/plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerTest.java
@@ -747,7 +747,8 @@ public class PluginManagerTest {
     public void resolveDependenciesFromManifestExceptionTest() {
         Plugin testPlugin = new Plugin("test", "latest", null, null);
         setTestUcJson();
-        assertThat(pm.resolveDependenciesFromManifest(testPlugin)).isEmpty();
+        assertThatThrownBy(() -> pm.resolveDependenciesFromManifest(testPlugin)).isInstanceOf(DownloadPluginException.class)
+        .hasMessageContaining("Unable to resolve dependencies");
     }
 
     @Test

--- a/plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerWiremockTest.java
+++ b/plugin-management-library/src/test/java/io/jenkins/tools/pluginmanager/impl/PluginManagerWiremockTest.java
@@ -7,6 +7,8 @@ import io.jenkins.tools.pluginmanager.config.Config;
 import io.jenkins.tools.pluginmanager.config.Credentials;
 import io.jenkins.tools.pluginmanager.config.Settings;
 import java.io.File;
+import java.io.IOException;
+import java.net.URL;
 import java.util.Collections;
 import org.junit.After;
 import org.junit.Before;
@@ -78,4 +80,12 @@ public class PluginManagerWiremockTest {
         int wireMockPort = archives.port(); // no credentials configured for this port
         assertThat(pm.downloadToFile("http://localhost:" + wireMockPort + "/protectedplugins/mailer/1.32/mailer.hpi", plugin, null)).isFalse();
     }
+
+    @Test
+    public void getJsonWithBasicAuth() throws IOException{
+        int wireMockPort = protectedArchives.port();
+        pm.setCm(new CacheManager(folder.newFolder().toPath(), false));
+        assertThat(pm.getJson(new URL("http://localhost:" + wireMockPort + "/update-center.json"), "cache-key")).isNotNull();
+    }
+
 }

--- a/plugin-management-library/src/test/resources/mappings/empty_update-centerjson.json
+++ b/plugin-management-library/src/test/resources/mappings/empty_update-centerjson.json
@@ -1,0 +1,19 @@
+{
+  "name" : "plugins_mailer_132_mailerhpi",
+  "request" : {
+    "url" : "/update-center.json",
+    "method" : "GET"
+  },
+  "response" : {
+    "status" : 200,
+    "body" : "{}",
+    "headers" : {
+      "Date" : "Sat, 25 Jul 2020 20:07:15 GMT",
+      "Server" : "Apache/2.4.29 (Ubuntu)",
+      "Last-Modified" : "Thu, 09 Apr 2020 10:51:31 GMT",
+      "ETag" : "\"1eda1-5a2d9664b503c\"",
+      "Accept-Ranges" : "bytes",
+      "Keep-Alive" : "timeout=3, max=99"
+    }
+  }
+}

